File: cycles/csycles/Shader.cs
/**
Copyright 2014 Robert McNeel and Associates

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/

using System.Collections.Generic;
using ccl.ShaderNodes;
using ccl.ShaderNodes.Sockets;

namespace ccl
{
	public class Shader
	{
		public enum ShaderType
		{
			Material,
			World
		}
		/// <summary>
		/// Get the ID for this shader
		/// </summary>
		public uint Id { get; private set; }
		private Client Client { get; set; }
		public ShaderType Type { get; set; }

		public OutputNode Output { get; set; }

		/// <summary>
		/// Create a new shader for client.
		/// </summary>
		/// <param name="clientId">Client ID for C[CS]ycles API.</param>
		/// <param name="type">The type of shader to create</param>
		public Shader(Client client, ShaderType type)
		{
			Client = client;
			Type = type;
			Id = CSycles.create_shader(Client.Id);
			Output = new OutputNode();
			AddNode(Output);
		}

		readonly List<ShaderNode> nodes = new List<ShaderNode>();
		/// <summary>
		/// Add a ShaderNode to the shader
		/// </summary>
		/// <param name="node">ShaderNode to add</param>
		public void AddNode(ShaderNode node)
		{
			if (node is OutputNode)
			{
				node.Id = CSycles.OUTPUT_SHADERNODE_ID;
				nodes.Add(node);
				return;
			}

			var nodeid = CSycles.add_shader_node(Client.Id, Id, node.Type);
			node.Id = nodeid;
			nodes.Add(node);

			/* set node attributes */
			if (node.inputs != null)
			{
				foreach (var socket in node.inputs.Sockets)
				{
					var float_socket = socket as FloatSocket;
					if (float_socket != null)
					{
						CSycles.shadernode_set_attribute_float(Client.Id, Id, node.Id, float_socket.Name, float_socket.Value);
					}
					var int_socket = socket as IntSocket;
					if (int_socket != null)
					{
						CSycles.shadernode_set_attribute_int(Client.Id, Id, node.Id, int_socket.Name, int_socket.Value);
					}
					var string_socket = socket as StringSocket;
					if (string_socket != null)
					{
						CSycles.shadernode_set_attribute_string(Client.Id, Id, node.Id, socket.Name, string_socket.Value);
					}
					var float4_socket = socket as Float4Socket;
					if (float4_socket != null)
					{
						CSycles.shadernode_set_attribute_vec(Client.Id, Id, node.Id, float4_socket.Name, float4_socket.Value);
					}
				}
			}

			/* set enums */
			switch (node.Type)
			{
				case ShaderNodeType.Math:
					CSycles.shadernode_set_enum(Client.Id, Id, node.Id, node.Type, ((MathNode)node).Operation.ToString().Replace('_', ' '));
					break;
				case ShaderNodeType.Refraction:
					CSycles.shadernode_set_enum(Client.Id, Id, node.Id, node.Type, ((RefractionBsdfNode)node).Distribution);
					break;
				case ShaderNodeType.Mix:
					CSycles.shadernode_set_enum(Client.Id, Id, node.Id, node.Type, ((MixNode)node).BlendType);
					break;
				case ShaderNodeType.Glossy:
					CSycles.shadernode_set_enum(Client.Id, Id, node.Id, node.Type, ((GlossyBsdfNode)node).Distribution);
					break;
				case ShaderNodeType.Glass:
					CSycles.shadernode_set_enum(Client.Id, Id, node.Id, node.Type, ((GlassBsdfNode)node).Distribution);
					break;
				case ShaderNodeType.SkyTexture:
					CSycles.shadernode_set_enum(Client.Id, Id, node.Id, node.Type, ((SkyTexture)node).SkyType);
					break;
				case ShaderNodeType.WaveTexture:
					CSycles.shadernode_set_enum(Client.Id, Id, node.Id, node.Type, ((WaveTexture)node).WaveType);
					break;
			}

			/* set direct member variables */
			switch (node.Type)
			{
				case ShaderNodeType.Math:
					CSycles.shadernode_set_member_bool(Client.Id, Id, node.Id, node.Type, "use_clamp", ((MathNode)node).UseClamp);
					break;
				case ShaderNodeType.Value:
					CSycles.shadernode_set_member_float(Client.Id, Id, node.Id, node.Type, "value", ((ValueNode)node).Value);
					break;
				case ShaderNodeType.Color:
					var val = ((ColorNode) node).Value;
					CSycles.shadernode_set_member_vec(Client.Id, Id, node.Id, node.Type, "value", val.x, val.y, val.z);
					break;
				case ShaderNodeType.ImageTexture:
					CSycles.shadernode_set_member_float(Client.Id, Id, node.Id, node.Type, "projection_blend", ((ImageTextureNode)node).ProjectionBlend);
					CSycles.shadernode_set_member_bool(Client.Id, Id, node.Id, node.Type, "is_linear", ((ImageTextureNode)node).IsLinear);
					CSycles.shadernode_set_member_bool(Client.Id, Id, node.Id, node.Type, "is_float", ((ImageTextureNode)node).IsFloat);
					break;
				case ShaderNodeType.BrickTexture:
					CSycles.shadernode_set_member_float(Client.Id, Id, node.Id, node.Type, "offset", ((BrickTexture)node).Offset);
					CSycles.shadernode_set_member_int(Client.Id, Id, node.Id, node.Type, "offset_frequency", ((BrickTexture)node).OffsetFrequency);
					CSycles.shadernode_set_member_float(Client.Id, Id, node.Id, node.Type, "squash", ((BrickTexture)node).Squash);
					CSycles.shadernode_set_member_int(Client.Id, Id, node.Id, node.Type, "squash_frequency", ((BrickTexture)node).SquashFrequency);
					break;
				case ShaderNodeType.SkyTexture:
					CSycles.shadernode_set_member_float(Client.Id, Id, node.Id, node.Type, "turbidity", ((SkyTexture)node).Turbidity);
					CSycles.shadernode_set_member_float(Client.Id, Id, node.Id, node.Type, "ground_albedo", ((SkyTexture)node).GroundAlbedo);
					var sd = ((SkyTexture) node).SunDirection;
					CSycles.shadernode_set_member_vec(Client.Id, Id, node.Id, node.Type, "sun_direction", sd.x, sd.y, sd.z);
					break;
			}
		}

		SocketBase lastNode { get; set; }

		/// <summary>
		/// Specify which ShaderNode is the last node for this Shader graph.
		/// Its output will be connected to the shader closure
		/// </summary>
		/// <param name="last">The ShaderNode that is last in the shader graph</param>
		/// <param name="outp">The output of the ShaderNode to connect to the final output</param>
		public void LastSocket(SocketBase last)
		{
			lastNode = last;
		}

		/// <summary>
		/// Push the defined shader graph to Cycles.
		/// </summary>
		public void FinalizeGraph()
		{
			foreach (var node in nodes)
			{
				if (node.inputs == null) continue;

				foreach(var socket in node.inputs.Sockets)
				{
					var from = socket.ConnectionFrom;
					if (from == null) continue;
					Connect(from.Parent, from.Name, node, socket.Name);
				}
			}

			//ConnectToOutput(lastNode.Parent, lastNode.Name);
		}

		private void Connect(ShaderNode from, string fromout, ShaderNode to, string toin)
		{
			if (nodes.Contains(from) && nodes.Contains(to))
			{
				CSycles.shader_connect_nodes(Client.Id, Id, from.Id, fromout, to.Id, toin);
			}
		}

		/** TODO: devise way to connect to different output types
		 * - Surface
		 * - Volume
		 * - World
		 */
		private void ConnectToOutput(ShaderNode from, string fromout)
		{
			if (nodes.Contains(from))
			{

				CSycles.shader_connect_nodes(Client.Id, Id, from.Id, fromout, CSycles.OUTPUT_SHADERNODE_ID, "Surface");
			}
		}

		/// <summary>
		/// Set the name of the Shader
		/// </summary>
		private string _name;
		public string Name
		{
			set
			{
				_name = value;
				CSycles.shader_set_name(Client.Id, Id, _name);
			}
			get
			{
				return _name;
			}
		}

		/// <summary>
		/// Set to true if Mis should be used.
		/// \todo find out what Mis actually means
		/// </summary>
		public bool UseMis
		{
			set
			{
				CSycles.shader_set_use_mis(Client.Id, Id, value);
			}
		}

		/// <summary>
		/// Set to true if this shader supports transparent shadows.
		/// </summary>
		public bool UseTransparentShadow
		{
			set
			{
				CSycles.shader_set_use_transparent_shadow(Client.Id, Id, value);
			}
		}

		/// <summary>
		/// Set to true if this shader is used for a heterogeneous volume.
		/// </summary>
		public bool HeterogeneousVolume
		{
			set
			{
				CSycles.shader_set_heterogeneous_volume(Client.Id, Id, value);
			}
		}
	}
}

File: cycles/ccycles/shader.cpp
/**
Copyright 2014 Robert McNeel and Associates

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/

#include "internal_types.h"

extern std::vector<CCScene> scenes;

std::vector<CCShader*> shaders;

std::vector<ccl::ImageManager::Image*> images;

void _cleanup_shaders()
{
	for (auto sh : shaders) {
		// just setting to nullptr, as scene disposal frees this memory.
		sh->graph = nullptr;
		sh->shader = nullptr;
		sh->scene_mapping.clear();
		delete sh;
	}
	shaders.clear();
}

void _cleanup_images()
{
	for (auto img : images) {
		if (img == nullptr) continue;

		delete [] img->builtin_data;
		img->builtin_data = nullptr;

		delete img;
	}
	images.clear();
}

/* Create a new shader.
 TODO: name for shader
*/
unsigned int cycles_create_shader(unsigned int client_id)
{
	CCShader* sh = new CCShader();
	sh->shader->graph = sh->graph;
	shaders.push_back(sh);

	return (unsigned int)(shaders.size() - 1);
}

/* Add shader to specified scene. */
unsigned int cycles_scene_add_shader(unsigned int client_id, unsigned int scene_id, unsigned int shader_id)
{
	SCENE_FIND(scene_id)
		auto sh = shaders[shader_id];
		sh->shader->tag_update(sce);
		sce->shaders.push_back(sh->shader);
		auto shid = (unsigned int)(sce->shaders.size() - 1);
		sh->scene_mapping.insert({ scene_id, shid });
		return shid;
	SCENE_FIND_END()

	return (unsigned int)(-1);
}

/* Get Cycles shader ID in specific scene. */
unsigned int cycles_scene_shader_id(unsigned int client_id, unsigned int scene_id, unsigned int shader_id)
{
	auto sh = shaders[shader_id];
	if (sh->scene_mapping.find(scene_id) != sh->scene_mapping.end()) {
		return sh->scene_mapping[scene_id];
	}

	return (unsigned int)(-1);
}

void cycles_shader_set_name(unsigned int client_id, unsigned int shader_id, const char* name)
{
	SHADER_SET(shader_id, string, name, name);
}

void cycles_shader_set_use_mis(unsigned int client_id, unsigned int shader_id, unsigned int use_mis)
{
	SHADER_SET(shader_id, bool, use_mis, use_mis == 1)
}

void cycles_shader_set_use_transparent_shadow(unsigned int client_id, unsigned int shader_id, unsigned int use_transparent_shadow)
{
	SHADER_SET(shader_id, bool, use_transparent_shadow, use_transparent_shadow == 1)
}

void cycles_shader_set_heterogeneous_volume(unsigned int client_id, unsigned int shader_id, unsigned int heterogeneous_volume)
{
	SHADER_SET(shader_id, bool, heterogeneous_volume, heterogeneous_volume == 1)
}

unsigned int cycles_add_shader_node(unsigned int client_id, unsigned int shader_id, shadernode_type shn_type)
{
	ccl::ShaderNode* node = nullptr;
	switch (shn_type) {
		case shadernode_type::BACKGROUND:
			node = new ccl::BackgroundNode();
			break;
		case shadernode_type::DIFFUSE:
			node = new ccl::DiffuseBsdfNode();
			break;
		case shadernode_type::ANISOTROPIC:
			node = new ccl::AnisotropicBsdfNode();
			break;
		case shadernode_type::TRANSLUCENT:
			node = new ccl::TranslucentBsdfNode();
			break;
		case shadernode_type::TRANSPARENT:
			node = new ccl::TransparentBsdfNode();
			break;
		case shadernode_type::VELVET:
			node = new ccl::VelvetBsdfNode();
			break;
		case shadernode_type::TOON:
			node = new ccl::ToonBsdfNode();
			break;
		case shadernode_type::GLOSSY:
			node = new ccl::GlossyBsdfNode();
			break;
		case shadernode_type::GLASS:
			node = new ccl::GlassBsdfNode();
			break;
		case shadernode_type::REFRACTION:
		{
			auto refrnode = new ccl::RefractionBsdfNode();
			refrnode->distribution = OpenImageIO::v1_3::ustring("Beckmann");
			node = dynamic_cast<ccl::ShaderNode *>(refrnode);
		}
			break;
		case shadernode_type::HAIR:
			node = new ccl::HairBsdfNode();
			break;
		case shadernode_type::EMISSION:
			node = new ccl::EmissionNode();
			break;
		case shadernode_type::AMBIENT_OCCLUSION:
			node = new ccl::AmbientOcclusionNode();
			break;
		case shadernode_type::ABSORPTION_VOLUME:
			node = new ccl::AbsorptionVolumeNode();
			break;
		case shadernode_type::SCATTER_VOLUME:
			node = new ccl::ScatterVolumeNode();
			break;
		case shadernode_type::SUBSURFACE_SCATTERING:
			node = new ccl::SubsurfaceScatteringNode();
			break;
		case shadernode_type::VALUE:
			node = new ccl::ValueNode();
			break;
		case shadernode_type::COLOR:
			node = new ccl::ColorNode();
			break;
		case shadernode_type::MIX_CLOSURE:
			node = new ccl::MixClosureNode();
			break;
		case shadernode_type::ADD_CLOSURE:
			node = new ccl::AddClosureNode();
			break;
		case shadernode_type::INVERT:
			node = new ccl::InvertNode();
			break;
		case shadernode_type::MIX:
			node = new ccl::MixNode();
			break;
		case shadernode_type::GAMMA:
			node = new ccl::GammaNode();
			break;
		case shadernode_type::WAVELENGTH:
			node = new ccl::WavelengthNode();
			break;
		case shadernode_type::BLACKBODY:
			node = new ccl::BlackbodyNode();
			break;
		case shadernode_type::CAMERA:
			node = new ccl::CameraNode();
			break;
		case shadernode_type::FRESNEL:
			node = new ccl::FresnelNode();
			break;
		case shadernode_type::COMBINE_XYZ:
			node = new ccl::CombineXYZNode();
			break;
		case shadernode_type::SEPARATE_XYZ:
			node = new ccl::SeparateXYZNode();
			break;
		case shadernode_type::MATH:
			node = new ccl::MathNode();
			break;
		case shadernode_type::IMAGE_TEXTURE:
			node = new ccl::ImageTextureNode();
			break;
		case shadernode_type::BRICK_TEXTURE:
			node = new ccl::BrickTextureNode();
			break;
		case shadernode_type::SKY_TEXTURE:
			node = new ccl::SkyTextureNode();
			break;
		case shadernode_type::CHECKER_TEXTURE:
			node = new ccl::CheckerTextureNode();
			break;
		case shadernode_type::NOISE_TEXTURE:
			node = new ccl::NoiseTextureNode();
			break;
		case shadernode_type::WAVE_TEXTURE:
			node = new ccl::WaveTextureNode();
			break;
		case shadernode_type::TEXTURE_COORDINATE:
			node = new ccl::TextureCoordinateNode();
			break;
	}

	if (node) {
		shaders[shader_id]->graph->add(node);
		return (unsigned int)(node->id);
	}
	else {
		return (unsigned int)-1;
	}
}

enum class attr_type {
	INT,
	FLOAT,
	FLOAT4,
	CHARP
};

struct attrunion {
	attr_type type;
	union {
		int i;
		float f;
		ccl::float4 f4;
		const char* cp;
	};
};

void shadernode_set_attribute(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, const char* attribute_name, attrunion v)
{
	auto attr = string(attribute_name);
	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			for (auto* inp : (*psh)->inputs) {
				auto inpname = string(inp->name);
				if (ccl::string_iequals(inpname, attribute_name)) {
					switch (v.type) {
					case attr_type::INT:
						inp->value.x = (float)v.i;
						logger.logit(client_id, "shader_id: ", shader_id, " -> shnode_id: ", shnode_id, " |> setting attribute: ", attribute_name, " to: ", v.i);
						break;
					case attr_type::FLOAT:
						inp->value.x = v.f;
						logger.logit(client_id, "shader_id: ", shader_id, " -> shnode_id: ", shnode_id, " |> setting attribute: ", attribute_name, " to: ", v.f);
						break;
					case attr_type::FLOAT4:
						inp->value.x = v.f4.x;
						inp->value.y = v.f4.y;
						inp->value.z = v.f4.z;
						logger.logit(client_id, "shader_id: ", shader_id, " -> shnode_id: ", shnode_id, " |> setting attribute: ", attribute_name, " to: ", v.f4.x, ",", v.f4.y, ",", v.f4.z);
						break;
					case attr_type::CHARP:
						inp->value_string = string(v.cp);
						logger.logit(client_id, "shader_id: ", shader_id, " -> shnode_id: ", shnode_id, " |> setting attribute: ", attribute_name, " to: ", v.cp);
						break;
					}
					return;
				}
			}
			break;
		}
		++psh;
	}
}

void _set_enum_val(unsigned int client_id, OpenImageIO::v1_3::ustring* str, ccl::ShaderEnum& enm, const OpenImageIO::v1_3::ustring val)
{
	if (enm.exists(val)) {
		*str = val;
	}
	else {
		logger.logit(client_id, "Unknown value ", val);
	}
}

/* TODO: add all enum possibilities.
 * Note that this particular API function won't deal yet very well with
 * nodes that have multiple enums.
 */
void cycles_shadernode_set_enum(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, shadernode_type shn_type, const char* value)
{
	auto val = OpenImageIO::v1_3::ustring(value);

	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			switch (shn_type) {
				case shadernode_type::MATH:
					{
						auto node = dynamic_cast<ccl::MathNode*>(*psh);
						_set_enum_val(client_id, &node->type, ccl::MathNode::type_enum, val);
					}
					break;
				case shadernode_type::MIX:
					{
						auto node = dynamic_cast<ccl::MixNode*>(*psh);
						_set_enum_val(client_id, &node->type, ccl::MixNode::type_enum, val);
					}
					break;
				case shadernode_type::REFRACTION:
					{
						auto node = dynamic_cast<ccl::RefractionBsdfNode*>(*psh);
						_set_enum_val(client_id, &node->distribution, ccl::RefractionBsdfNode::distribution_enum, val);
					}
					break;
				case shadernode_type::GLOSSY:
					{
						auto node = dynamic_cast<ccl::GlossyBsdfNode*>(*psh);
						_set_enum_val(client_id, &node->distribution, ccl::GlossyBsdfNode::distribution_enum, val);
					}
					break;
				case shadernode_type::GLASS:
					{
						auto node = dynamic_cast<ccl::GlassBsdfNode*>(*psh);
						_set_enum_val(client_id, &node->distribution, ccl::GlassBsdfNode::distribution_enum, val);
					}
					break;
				case shadernode_type::WAVE_TEXTURE:
					{
						auto node = dynamic_cast<ccl::WaveTextureNode*>(*psh);
						_set_enum_val(client_id, &node->type, ccl::WaveTextureNode::type_enum, val);
					}
					break;
				case shadernode_type::SKY_TEXTURE:
					{
						auto node = dynamic_cast<ccl::SkyTextureNode*>(*psh);
						_set_enum_val(client_id, &node->type, ccl::SkyTextureNode::type_enum, val);
					}
					break;
			}
			break;
		}
		++psh;
	}
}

void cycles_shadernode_set_member_float_img(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, shadernode_type shn_type, const char* member_name, const char* img_name, float* img, unsigned int width, unsigned int height, unsigned int depth, unsigned int channels)
{
	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	auto mname = string(member_name);
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			switch (shn_type) {
				case shadernode_type::IMAGE_TEXTURE:
					{
						auto nimg = new ccl::ImageManager::Image();
						auto imgdata = new float[width*height*channels];
						memcpy(imgdata, img, sizeof(float)*width*height*channels);
						nimg->builtin_data = imgdata;
						nimg->filename = string(img_name);
						nimg->width = (int)width;
						nimg->height = (int)height;
						nimg->depth = (int)depth;
						nimg->channels = (int)channels;
						nimg->is_float = true;
						images.push_back(nimg);
						auto imtex = dynamic_cast<ccl::ImageTextureNode*>(*psh);
						imtex->builtin_data = nimg;
						imtex->is_float = true;
						imtex->is_linear = true;
						imtex->interpolation = ccl::InterpolationType::INTERPOLATION_LINEAR;
					}	
					break;
			}
		}
		++psh;
	}
}
void cycles_shadernode_set_member_byte_img(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, shadernode_type shn_type, const char* member_name, const char* img_name, unsigned char* img, unsigned int width, unsigned int height, unsigned int depth, unsigned int channels)
{
	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	auto mname = string(member_name);
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			switch (shn_type) {
				case shadernode_type::IMAGE_TEXTURE:
					{
						auto nimg = new ccl::ImageManager::Image();
						auto imgdata = new ccl::uchar[width*height*channels];
						memcpy(imgdata, img, sizeof(unsigned char)*width*height*channels);
						nimg->builtin_data = imgdata;
						nimg->filename = string(img_name);
						nimg->width = (int)width;
						nimg->height = (int)height;
						nimg->depth = (int)depth;
						nimg->channels = (int)channels;
						nimg->is_float = false;
						images.push_back(nimg);
						auto imtex = dynamic_cast<ccl::ImageTextureNode*>(*psh);
						imtex->builtin_data = nimg;
						imtex->is_float = false;
						imtex->is_linear = true;
						imtex->interpolation = ccl::InterpolationType::INTERPOLATION_LINEAR;
					}
					break;
			}
		}
		++psh;
	}
}

void cycles_shadernode_set_member_bool(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, shadernode_type shn_type, const char* member_name, bool value)
{
	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	auto mname = string(member_name);
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			switch (shn_type) {
				case shadernode_type::MATH:
					{
						auto mnode = dynamic_cast<ccl::MathNode*>(*psh);
						mnode->use_clamp = value;
					}
					break;
				case shadernode_type::IMAGE_TEXTURE:
					{
						auto imtexnode = dynamic_cast<ccl::ImageTextureNode*>(*psh);
						if (mname == "is_linear") {
							imtexnode->is_linear = value;
						}
						else if (mname == "if_float") {
							imtexnode->is_float = value;
						}
					}
					break;
			}
		}
	++psh;
	}

}

void cycles_shadernode_set_member_int(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, shadernode_type shn_type, const char* member_name, int value)
{
	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	auto mname = string(member_name);
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			switch (shn_type) {
				case shadernode_type::BRICK_TEXTURE:
					{
						auto bricknode = dynamic_cast<ccl::BrickTextureNode*>(*psh);
						if (mname == "offset_frequency")
							bricknode->offset_frequency = value;
						else if (mname == "squash_frequency")
							bricknode->squash_frequency = value;
					}
					break;
			}
		}
		++psh;
	}
}


void cycles_shadernode_set_member_float(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, shadernode_type shn_type, const char* member_name, float value)
{
	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	auto mname = string(member_name);
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			switch (shn_type) {
				case shadernode_type::VALUE:
					{
						auto valuenode = dynamic_cast<ccl::ValueNode*>(*psh);
						valuenode->value = value;
					}
					break;
				case shadernode_type::IMAGE_TEXTURE:
					{
						auto imtexnode = dynamic_cast<ccl::ImageTextureNode*>(*psh);
						if (mname == "projection_blend") {
							imtexnode->projection_blend = value;
						}
					}
					break;
				case shadernode_type::BRICK_TEXTURE:
					{
						auto bricknode = dynamic_cast<ccl::BrickTextureNode*>(*psh);
						if (mname == "offset")
							bricknode->offset = value;
						else if (mname == "squash")
							bricknode->squash = value;
					}
					break;
				case shadernode_type::SKY_TEXTURE:
					{
						auto skynode = dynamic_cast<ccl::SkyTextureNode*>(*psh);
						if (mname == "turbidity")
							skynode->turbidity = value;
						else if (mname == "ground_albedo")
							skynode->ground_albedo = value;
					}
					break;
			}
		}
		++psh;
	}
}

void cycles_shadernode_set_member_vec(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, shadernode_type shn_type, const char* member_name, float x, float y, float z)
{
	auto& sh = shaders[shader_id];
	auto psh = sh->graph->nodes.begin();
	while (psh != sh->graph->nodes.end())
	{
		if ((*psh)->id == shnode_id) {
			switch (shn_type) {
			case shadernode_type::COLOR:
					{
						auto colnode = dynamic_cast<ccl::ColorNode*>(*psh);
						colnode->value.x = x;
						colnode->value.y = y;
						colnode->value.z = z;
					}
					break;
			case shadernode_type::SKY_TEXTURE:
					{
						auto sunnode = dynamic_cast<ccl::SkyTextureNode*>(*psh);
						sunnode->sun_direction.x = x;
						sunnode->sun_direction.y = y;
						sunnode->sun_direction.z = z;
					}
					break;
			}
		}
		++psh;
	}
}

/*
Set an integer attribute with given name to value. shader_id is the global shader ID.
*/
void cycles_shadernode_set_attribute_int(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, const char* attribute_name, int value)
{
	attrunion v;
	v.type = attr_type::INT;
	v.i = value;
	shadernode_set_attribute(client_id, shader_id, shnode_id, attribute_name, v);
}

/*
Set a float attribute with given name to value. shader_id is the global shader ID.
*/
void cycles_shadernode_set_attribute_float(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, const char* attribute_name, float value)
{
	attrunion v;
	v.type = attr_type::FLOAT;
	v.f = value;
	shadernode_set_attribute(client_id, shader_id, shnode_id, attribute_name, v);
}

/*
Set a vector of floats attribute with given name to x, y and z. shader_id is the global shader ID.
*/
void cycles_shadernode_set_attribute_vec(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, const char* attribute_name, float x, float y, float z)
{
	attrunion v;
	v.type = attr_type::FLOAT4;
	v.f4.x = x;
	v.f4.y = y;
	v.f4.z = z;
	shadernode_set_attribute(client_id, shader_id, shnode_id, attribute_name, v);
}

/*
Set a string attribute with given name to value. shader_id is the global shader ID.
*/
void cycles_shadernode_set_attribute_string(unsigned int client_id, unsigned int shader_id, unsigned int shnode_id, const char* attribute_name, const char* value)
{
	attrunion v;
	v.type = attr_type::CHARP;
	v.cp = value;
	shadernode_set_attribute(client_id, shader_id, shnode_id, attribute_name, v);
}

void cycles_shader_connect_nodes(unsigned int client_id, unsigned int shader_id, unsigned int from_id, const char* from, unsigned int to_id, const char* to)
{
	auto& sh = shaders[shader_id];
	auto shfrom = sh->graph->nodes.begin();
	auto shfrom_end = sh->graph->nodes.end();
	auto shto = sh->graph->nodes.begin();
	auto shto_end = sh->graph->nodes.end();

	while (shfrom != shfrom_end) {
		if ((*shfrom)->id == from_id) {
			break;
		}
		++shfrom;
	}

	while (shto != shto_end) {
		if ((*shto)->id == to_id) {
			break;
		}
		++shto;
	}

	if (shfrom == shfrom_end || shto == shto_end) {
		return; // TODO: figure out what to do on errors like this
	}
	logger.logit(client_id, "Shader ", shader_id, " :: ", from_id, ":", from, " -> ", to_id, ":", to);

	sh->graph->connect((*shfrom)->output(from), (*shto)->input(to));
}

